src/charm.py

```diff
@@ -111,14 +102,44 @@ def generate_worker_layer(self, worker: Worker) -> Layer:
             }
         )
 
+    def _add_juju_topolgy(self):
+        """Modify the config file to add juju topology for
+        `metrics-generator` metrics before running the pebble service.
+        """
+        tempo_container = self.unit.get_container("tempo")
+        if not tempo_container.exists(CONFIG_FILE):
+            logger.debug("Skipping adding juju topology. Config file does not exist yet.")
+            return
+
+        raw_data = tempo_container.pull(CONFIG_FILE).read()
+        data = yaml.safe_load(raw_data)
+        # if `metrics_generator` doesn't exist in config,
+        # then it is not enabled, so no point of adding juju topology.
+        if "metrics_generator" in data:
+            if "registry" not in data["metrics_generator"]:
+                data["metrics_generator"]["registry"] = {}
+            data["metrics_generator"]["registry"]["external_labels"] = dict(
+                self.worker.cluster.juju_topology.as_dict()
+            )
+
+            # write new data
+            tempo_container.push(CONFIG_FILE, yaml.safe_dump(data))
+
     def _on_collect_status(self, e: CollectStatusEvent):
-        # add Tempo worker custom blocking conditions
-        if self.worker.roles and self.worker.roles[0] not in self._valid_roles:
```

Comment:
we probably want to keep this blocking condition. Why remove it?

Reply:
The `self.worker.roles` is already covered in [abstract worker](https://github.com/canonical/cos-lib/blob/main/src/cosl/coordinated_workers/worker.py#L301). The `valid_roles` is not needed anymore after we've changed the roles config options to be `role-X: boolean`. 
I changed it with another check to block if multiple roles are enabled.